From b36dfb3b6a2d6a2e6c557217e39f88919cc5a017 Mon Sep 17 00:00:00 2001
From: Amith A <quic_amitajit@quicinc.com>
Date: Mon, 25 Mar 2024 19:19:28 +0530
Subject: [PATCH] wifi: ath12k: Set need_ptk_4_way flag to false for FT roaming

In case of FT roaming, since there is no EAPOL handshake, FW should be
notified not to wait till EAPOL handshake is completed to send data
packets. Set need_ptk_4_way flag to false to inform FW not to stall the
data packets till handshake completion in case of FT roaming.

Signed-off-by: Amith A <quic_amitajit@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 740af52..ee03756 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -2223,7 +2223,11 @@ static void ath12k_peer_assoc_h_crypto(struct ath12k *ar,
 			       IEEE80211_BSS_TYPE_ANY, IEEE80211_PRIVACY_ANY);
 
 	if (ahvif->rsnie_present || ahvif->wpaie_present) {
-		arg->need_ptk_4_way = true;
+		if (sta->ft_auth)
+			arg->need_ptk_4_way = false;
+		else
+			arg->need_ptk_4_way = true;
+
 		if (ahvif->wpaie_present)
 			arg->need_gtk_2_way = true;
 	} else if (bss) {
@@ -2244,7 +2248,10 @@ static void ath12k_peer_assoc_h_crypto(struct ath12k *ar,
 	if (ar->supports_6ghz || rsnie || wpaie) {
 		ath12k_dbg(ar->ab, ATH12K_DBG_WMI,
 			   "%s: rsn ie found\n", __func__);
-		arg->need_ptk_4_way = true;
+		if (sta->ft_auth)
+			arg->need_ptk_4_way = false;
+		else
+			arg->need_ptk_4_way = true;
 	}
 
 	if (wpaie) {
-- 
2.34.1

